lib/engine-universal/src/executable.rs

```diff
@@ -0,0 +1,286 @@
+/// supported compilers. This type falls in-between a module and [`Artifact`](crate::Artifact).
+#[derive(MemoryUsage, rkyv::Archive, rkyv::Deserialize, rkyv::Serialize)]
+pub struct UniversalExecutable {
+    pub(crate) function_bodies: PrimaryMap<LocalFunctionIndex, FunctionBody>,
+    pub(crate) function_relocations: PrimaryMap<LocalFunctionIndex, Vec<Relocation>>,
+    pub(crate) function_jt_offsets: PrimaryMap<LocalFunctionIndex, JumpTableOffsets>,
+    pub(crate) function_frame_info: PrimaryMap<LocalFunctionIndex, CompiledFunctionFrameInfo>,
+    pub(crate) function_call_trampolines: PrimaryMap<SignatureIndex, FunctionBody>,
+    pub(crate) dynamic_function_trampolines: PrimaryMap<FunctionIndex, FunctionBody>,
+    pub(crate) custom_sections: PrimaryMap<SectionIndex, CustomSection>,
+    pub(crate) custom_section_relocations: PrimaryMap<SectionIndex, Vec<Relocation>>,
+    // The section indices corresponding to the Dwarf debug info
+    pub(crate) debug: Option<Dwarf>,
+    // the Trampoline for Arm arch
+    pub(crate) trampolines: Option<TrampolinesSection>,
+    pub(crate) compile_info: CompileModuleInfo,
+    pub(crate) data_initializers: Vec<OwnedDataInitializer>,
+    pub(crate) cpu_features: u64,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ExecutableSerializeError {
+    #[error("could not serialize the executable data")]
+    Executable(
+        #[source]
+        CompositeSerializerError<
+            std::convert::Infallible,
+            AllocScratchError,
+            SharedSerializeMapError,
+        >,
+    ),
+}
+
+// SAFETY: the pointers in `rkyv::AllocScratchError` are present there for display purposes – this
+// type does not expose any mechanism to access or operate on the pointers, unless they are
+// accessed directly by the user.
+unsafe impl Send for ExecutableSerializeError {}
+unsafe impl Sync for ExecutableSerializeError {}
+
+impl wasmer_engine::Executable for UniversalExecutable {
+    fn load(
+        &self,
+        engine: &(dyn Engine + 'static),
+    ) -> Result<std::sync::Arc<dyn Artifact>, CompileError> {
+        engine
+            .downcast_ref::<crate::UniversalEngine>()
+            .ok_or_else(|| CompileError::Codegen("can't downcast TODO FIXME".into()))?
+            .load_universal_executable(self)
+            .map(|a| Arc::new(a) as _)
+    }
+
+    fn features(&self) -> Features {
+        self.compile_info.features.clone()
+    }
+
+    fn cpu_features(&self) -> EnumSet<CpuFeature> {
+        EnumSet::from_u64(self.cpu_features)
+    }
+
+    fn serialize(&self) -> Result<Vec<u8>, Box<(dyn std::error::Error + Send + Sync + 'static)>> {
+        // The format is as thus:
+        //
+        // HEADER
+        // RKYV PAYLOAD
+        // RKYV POSITION
+        //
+        // It is expected that any framing for message length is handled by the caller.
+        let mut out = Vec::with_capacity(32);
+        out.extend(&MAGIC_HEADER);
```

Comment:
`MAGIC_HEADER.to_vec()`, maybe? We also are able to pre-allocate all the space we need ahead of time here.